### ISSUE TITLE
GPORCA: dummy selectivity for equality predicate with enum arguments

### ISSUE THREAD

**Issue by maksm90** [greenplum-db/gpdb#11520]
Hello everyone!

One of our (Arenadata) customers has faced with issue of far from optimal cardinality estimation for selection with equality predicate containing enum-typed arguments. For any enum constant in predicate (even for very sparse or frequent) ORCA assigns to it a dummy selectivity that much worse than estimation from PostgreSQL optimizer.

### Step to reproduce the behavior

```sql
create type custom_enum as enum ('attending_value', 'missing_value');
create table test(t custom_enum);
insert into test select 'attending_value' from generate_series(1, 1000);
analyze test;

# check whether MCV/MCF for enum-typed column is not empty
select most_common_vals, most_common_freqs from pg_stats where tablename = 'test' and attname = 't';
 most_common_vals  | most_common_freqs 
-------------------+-------------------
 {attending_value} | {1}
```

### Expected behavior

```sql
# PostgreSQL optimizer computes selectivity of predicate based on MCV/MCF values
set optimizer to off;

explain select * from test where t = 'attending_value';
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..14.50 rows=1000 width=4)
   ->  Seq Scan on test  (cost=0.00..14.50 rows=1000 width=4)
         Filter: (t = 'attending_value'::custom_enum)
 Optimizer: Postgres query optimizer
 
 explain select * from test where t = 'missing_value';
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..14.50 rows=1 width=4)
   ->  Seq Scan on test  (cost=0.00..14.50 rows=1 width=4)
         Filter: (t = 'missing_value'::custom_enum)
 Optimizer: Postgres query optimizer
```

### Actual behavior

```sql
# ORCA optimizer assigns dummy selectivity to predicate for both constant values
set optimizer to on;

explain select * from test where t = 'attending_value';
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.07 rows=400 width=4)
   ->  Seq Scan on test  (cost=0.00..431.05 rows=*400* width=4)
         Filter: (t = 'attending_value'::custom_enum)
 Optimizer: Pivotal Optimizer (GPORCA)
 
explain select * from test where t = 'missing_value';
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.07 rows=400 width=4)
   ->  Seq Scan on test  (cost=0.00..431.05 rows=*400* width=4)
         Filter: (t = 'missing_value'::custom_enum)
 Optimizer: Pivotal Optimizer (GPORCA)
```
&nbsp;
As far as I could find out the similar type of arguments that is considered as custom (user defined) doesn't allow to be mapped to primitive one (LINT or DOUBLE) to be further located on histogram line. Therefore, ORCA [doesn't get MCV/MCF values from MD provider](https://github.com/greenplum-db/gpdb/blob/6.14.0/src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp#L2834-L2841) and treats predicate as one with uncomoarable statistics (the [case 2](https://github.com/greenplum-db/gpdb/blob/6.14.0/src/backend/gporca/libnaucrates/src/statistics/CStatsPredUtils.cpp#L303-L312) inside conditional branch). The second factor causes assignment of dummy selectivity to the predicate via [set up of scale_factor](https://github.com/greenplum-db/gpdb/blob/6.14.0/src/backend/gporca/libnaucrates/src/statistics/CFilterStatsProcessor.cpp#L339-L340) that is inverse from selectivity.

I see two ways to fix current issue:

1. Add mapping enums to LINT values represented as OIDs of enum scalars. But this requires to account order of enum scalars that is specified by `enumsortorder` column of corresponding `pg_enum` entries.
2. For any types not mapped to LINT or DOUBLE, delegate selectivity execution to PostgreSQL estimation function that is defined in `pg_operator->oprrest` column of operator participating inside predicate. For equality of enums this function (`eqsel`) computes selectivity [based on](https://github.com/greenplum-db/gpdb/blob/6.14.0/src/backend/utils/adt/selfuncs.c#L286-L371) MCV and MCF values (the basic variant of such computation):
    * for constant inside MCV the corresponding frequency in MCF is returned
    * for other case the following formula works:
      ```
      selectivity = (1 - sum(MCF) - null_freq) / (num_distinct - len(MCV))
      ```
      if `num_distinct != len(MCV)`, else
      ``` 
      selectivity = (1 - sum(MCF) - null_freq)
      ```
      resulting value has not to exceed a minimal frequency in MCF list and assigns to it in opposite case.

This second point also incorporates selectivity estimation for all custom types (that is not mapped to LINT or DOUBLE), not only for enums as in the first option.

Any thoughts, suggestions?

### Greenplum version or build

master and below

**Comment by d** [comment]
> 2\. For any types not mapped to LINT or DOUBLE, delegate selectivity execution to PostgreSQL estimation function that is defined in `pg_operator->oprrest` column of operator participating inside predicate. For equality of enums this function (`eqsel`) computes selectivity [based on](https://github.com/greenplum-db/gpdb/blob/6.14.0/src/backend/utils/adt/selfuncs.c#L286-L371) MCV and MCF values (the basic variant of such computation):
>     
>     * for constant inside MCV the corresponding frequency in MCF is returned
>     * for other case the following formula works:
>       ```
>       selectivity = (1 - sum(MCF) - null_freq) / (num_distinct - len(MCV))
>       ```
>       
>       
>       if `num_distinct != len(MCV)`, else
>       ```
>       selectivity = (1 - sum(MCF) - null_freq)
>       ```
>       
>       
>       resulting value has not to exceed a minimal frequency in MCF list and assigns to it in opposite case.
> 
> This second point also incorporates selectivity estimation for all custom types (that is not mapped to LINT or DOUBLE), not only for enums as in the first option.
> 
> Any thoughts, suggestions?

I'll be very excited if we can explore a solution that looks like that! Looking a bit further, this might pave the way for ORCA to leverage the more advanced planner support functions (like rewriting `a LIKE 'abc%'` into a `BETWEEN` clause)

**Comment by maksm90** [comment]
> Looking a bit further, this might pave the way for ORCA to leverage the more advanced planner support functions (like rewriting a LIKE 'abc%' into a BETWEEN clause)

Yeah, in perspective, any unsolved cases in the context of selectivity estimation we would able to delegate to postgres optimizer.
